executor/src/witgen/query_processor.rs

```diff
@@ -56,50 +57,82 @@ where
-                .constant_value()
-                .ok_or(IncompleteCause::NonConstantQueryMatchScrutinee)?;
-            let expr = arms
+    fn interpolate_query(
+        &self,
+        query: &'a Expression<T>,
+        rows: &RowPair<T>,
+    ) -> Result<String, IncompleteCause<&'a PolynomialReference>> {
+        // TODO combine that with the constant evaluator and the commit evaluator...
+        match query {
+            Expression::Tuple(items) => Ok(items
                 .iter()
-                .find_map(|MatchArm { pattern, value }| {
-                    (match pattern {
-                        MatchPattern::CatchAll => true,
-                        MatchPattern::Pattern(pattern) => {
-                            rows.evaluate(pattern).unwrap().constant_value() == Some(v)
-                        }
+                .map(|i| self.interpolate_query(i, rows))
+                .collect::<Result<Vec<_>, _>>()?
+                .join(", ")),
+            Expression::String(s) => Ok(format!(
+                "\"{}\"",
+                s.replace('\\', "\\\\").replace('"', "\\\"")
+            )),
+            Expression::MatchExpression(scrutinee, arms) => {
+                let v = self.evaluate_expression(scrutinee, rows)?;
+                let expr = arms
+                    .iter()
+                    .find_map(|MatchArm { pattern, value }| {
+                        (match pattern {
+                            MatchPattern::CatchAll => true,
+                            MatchPattern::Pattern(pattern) => {
+                                self.evaluate_expression(pattern, rows) == Ok(v)
+                            }
+                        })
+                        .then_some(value)
                     })
-                    .then_some(value)
-                })
-                .ok_or(IncompleteCause::NoMatchArmFound)?;
-            interpolate_query(expr, rows)
+                    .ok_or(IncompleteCause::NoMatchArmFound)?;
+                self.interpolate_query(expr, rows)
+            }
+            _ => self.evaluate_expression(query, rows).map(|v| v.to_string()),
+        }
+    }
+
+    fn evaluate_expression(
+        &self,
+        expr: &'a Expression<T>,
+        rows: &RowPair<T>,
+    ) -> Result<T, IncompleteCause<&'a PolynomialReference>> {
+        match expr {
+            Expression::Number(n) => Ok(*n),
+            Expression::BinaryOperation(left, op, right) => Ok(evaluate_binary_operation(
+                self.evaluate_expression(left, rows)?,
+                *op,
+                self.evaluate_expression(right, rows)?,
+            )),
+            Expression::UnaryOperation(op, expr) => Ok(evaluate_unary_operation(
+                *op,
+                self.evaluate_expression(expr, rows)?,
+            )),
+            Expression::Reference(Reference::LocalVar(i, _name)) => {
+                assert!(*i == 0);
+                Ok(rows.current_row_index.into())
```

Comment:
Nice! I think doing this recursively fixes a few cases that previously wouldn't be evaluated correctly.